updates.jenkins.io.tf

```diff
@@ -3,3 +3,28 @@ resource "cloudflare_r2_bucket" "westeurope_updates_jenkins_io" {
   name       = "westeurope-updates-jenkins-io"
   location   = "WEUR"
 }
+
+# Keeping this main zone imported as data only so we don't risk destroying it (the sponsoring is attached to it)
+data "cloudflare_zone" "cloudflare_jenkins_io" {
+  account_id = local.account_id["jenkins-infra-team"]
+  name       = "cloudflare.jenkins.io"
+}
+
+## West Europe
+resource "cloudflare_record" "westeurope" {
+  for_each = toset(data.cloudflare_zone.cloudflare_jenkins_io.name_servers)
```

Comment:
this doesn't seem right to me.

normally you would create a zone and then in the parent zone you would create an NS record with values for each of the child zones nameservers.

(I've never used cloudfare though)

Reply:
+1 with @timja , it looks like a mistake.

As per our discussion, the goal is still what we understood:

- For each R2 bucket we want public, we need a public zone.
- Each public zone is a child of the top-level zone
- The top-level zone has the delegation from Azure

It means that the NS records for the child zone must point to the child's zone name servers (not the parents) .

```suggestion
  for_each = toset(cloudflare_zone.westeurope_cloudflare_jenkins_io.name_servers)
```

Reply:
i would also reorder the file so that it goes data source, zone and then records

Reply:
To manually create the existing `europe` (test) zone I had to first add two NS records to the `cloudflare.jenkins.io` zone pointing to `cloudflare.jenkins.io` name servers:

<img width="1010" alt="image" src="https://github.com/jenkins-infra/cloudflare/assets/91831478/cd4f6c74-6d83-4248-893f-bd5c7b3f493a">

Then I was able to create/add to the interface the `europe.cloudflare.jenkins.io` website/zone, which I was then able to use as custom public domain for the `western-europe` (test) R2 bucket:

<img width="964" alt="image" src="https://github.com/jenkins-infra/cloudflare/assets/91831478/01bfc0e4-53d0-4e3f-a455-d098a0e9232a">

@dduportal I don't think your suggestion is possible, as I would need the zone to be created to get its name servers.

Reply:
yes, that is what terraform is made for. It will detect the dependency and will manage to first create the child zone, and then the records.

Reply:
Ok, let's see what it does with your suggestion.

Reply:
Error from build logs with your suggestion:

> The "for_each" value depends on resource attributes that cannot be determined until apply, so Terraform cannot predict how many instances will be created. To work around this, use the -target argument to first apply only the resources that the for_each depends on.

Reply:
that's annoying 😢.

Maybe try something like this https://stackoverflow.com/a/76692461/4951015 so that the count isn't hardcoded?

Reply:
@timja what would be the reference? `cloudflare.jenkins.io` name servers count?

Something like `count = length(data.cloudflare_jenkins_io.name_servers)`?

Reply:
> @timja what would be the reference? `cloudflare.jenkins.io` name servers count?
> 
> Something like `count = length(data.cloudflare_jenkins_io.name_servers)`?

Yep, that is the suggestion I was gonna make. But not sure it will work either: I believe count interpolation is done at the same time as the `for_each` in Terraform 1.x (it was different with 0.x).

I don't see the hardcoded value as a problem though.

Another path: can we define a mutli-valued NS record through terraform (Cloudflare Ui does not allow it but maybe API/terraform works under the hood), like it's done with Azure.

Reply:
> Another path: can we define a mutli-valued NS record through terraform (Cloudflare Ui does not allow it but maybe API/terraform works under the hood), like it's done with Azure.

I haven't found a way to do it via the provider: https://registry.terraform.io/providers/cloudflare/cloudflare/latest/docs/resources/record